component/workspace-manager.js

```diff
@@ -8,28 +8,52 @@ const templateRegistry = require('./template-registry');
  */
 const Manager = class Manager {
   constructor() {
+    this.listOfWorkspaces = {};
+    this.index = 0;
+    this.folderMap = {};
   }
   createWorkspace(dir) {
     if (this.workspace && this.workspace.getDirectory() === dir) {
       return this.workspace;
     }
-    this.workspace = new Workspace(dir);
-    this.workspace.addDomain('Facet');
-    this.workspace.addDomain('FacetConfig');
-    this.workspace.addDomain('DataSource');
-    this.workspace.addDomain('MiddlewarePhase');
-    this.workspace.addDomain('Middleware');
-    this.workspace.addDomain('ModelConfig');
-    this.workspace.addDomain('ModelDefinition');
-    this.workspace.addDomain('ModelProperty');
-    this.workspace.addDomain('ModelRelation');
-    this.workspace.addDomain('PackageDefinition');
-    this.initMiddleware(this.workspace);
-    this.workspace.addBuiltInModel('User');
-    return this.workspace;
-  }
-  getWorkspace() {
-    return this.workspace;
+    const workspace = new Workspace(dir);
+    workspace.addDomain('Facet');
+    workspace.addDomain('FacetConfig');
+    workspace.addDomain('DataSource');
+    workspace.addDomain('MiddlewarePhase');
+    workspace.addDomain('Middleware');
+    workspace.addDomain('ModelConfig');
+    workspace.addDomain('ModelDefinition');
+    workspace.addDomain('ModelProperty');
+    workspace.addDomain('ModelRelation');
+    workspace.addDomain('PackageDefinition');
+    this.initMiddleware(workspace);
+    workspace.addBuiltInModel('User');
+    workspace.setId(this.getWorkspaceId());
+    this.listOfWorkspaces[workspace.getId()] = workspace;
+    this.folderMap[dir] = workspace.getId();
+    if (this.index === 1) this.defaultWorkspace = workspace;
+    return workspace;
+  }
+  getWorkspaceId() {
+    this.index++;
+    const prefix = '0000';
+    let id = '' + this.index;
+    return prefix.substring(0, (prefix.length - id.length)).concat(id);
+  }
+  getWorkspace(id) {
+    if (id) return this.listOfWorkspaces[id];
+    else return this.defaultWorkspace;
+  }
+  getWorkspaceByFolder(dir) {
+    if (this.folderMap[dir]) return this.listOfWorkspaces[this.folderMap[dir]];
+  }
+  deleteWorkspace(id) {
+    if (this.listOfWorkspaces[id]) {
+      const dir = this.listOfWorkspaces[id].getDirectory();
+      delete this.folderMap[dir];
+      delete this.listOfWorkspaces[id];
+    }
```

Comment:
delete will be performed only if the id is available. it is upto the caller to validate and handle the conditions